File: dotnet/src/Microsoft.AutoGen/Core.Grpc/IAgentMessageSerializer.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// IAgentMessageSerializer.cs

namespace Microsoft.AutoGen.Core.Grpc;
/// <summary>
/// Interface for serializing and deserializing agent messages.
/// </summary>
public interface IAgentMessageSerializer
{
    /// <summary>
    /// Serialize an agent message.
    /// </summary>
    /// <param name="message">The message to serialize.</param>
    /// <returns>The serialized message.</returns>
    Google.Protobuf.WellKnownTypes.Any Serialize(object message);

    /// <summary>
    /// Deserialize an agent message.
    /// </summary>
    /// <param name="message">The message to deserialize.</param>
    /// <returns>The deserialized message.</returns>
    object Deserialize(Google.Protobuf.WellKnownTypes.Any message);
}

File: typescript/src/core.grpc/IAgentMessageSerializer.ts
import { Message } from 'google-protobuf';

/**
 * Interface for serializing messages between agent runtime and protobuf format.
 */
export interface IAgentMessageSerializer {
    /**
     * Serializes a message to protobuf format.
     * @param message The message to serialize
     * @returns The protobuf message
     */
    serialize(message: unknown): Message;

    /**
     * Deserializes a protobuf message.
     * @param protoMessage The protobuf message to deserialize
     * @returns The deserialized message
     */
    deserialize(protoMessage: Message): unknown;

    /**
     * Gets the protobuf message type for a given message.
     * @param message The message to get type for
     * @returns The protobuf message type name
     */
    getProtoMessageType(message: unknown): string;

    /**
     * Gets whether this serializer can handle a given message type.
     * @param messageType The message type name to check
     * @returns True if this serializer can handle the message type
     */
    canHandle(messageType: string): boolean;
}
